src/ObjectUtils/CoffFile.cpp

```diff
@@ -258,34 +315,131 @@ ErrorMessageOr<ModuleSymbols> CoffFileImpl::LoadDebugSymbols() {
+// Even llvm-objdump has to do that:
+// https://github.com/llvm-mirror/llvm/blob/master/tools/llvm-objdump/COFFDump.cpp
+ErrorMessageOr<llvm::ArrayRef<llvm::Win64EH::RuntimeFunction>> CoffFileImpl::GetRuntimeFunctions() {
+  const llvm::object::data_directory* exception_table_data_dir =
+      object_file_->getDataDirectory(llvm::COFF::EXCEPTION_TABLE);
+  if (exception_table_data_dir == nullptr) {
+    return ErrorMessage{"Unable to read Exception Table: No corresponding Data Directory."};
+  }
+
+  llvm::ArrayRef<uint8_t> exception_table_bytes;
+  if (object_file_->getRvaAndSizeAsBytes(exception_table_data_dir->RelativeVirtualAddress,
+                                         exception_table_data_dir->Size, exception_table_bytes)) {
+    return ErrorMessage{"Unable to read Exception Table."};
+  }
+
+  if (exception_table_bytes.size() % sizeof(llvm::Win64EH::RuntimeFunction) != 0) {
+    return ErrorMessage{"Unable to read Exception Table: Unexpected size."};
+  }
+
+  return llvm::ArrayRef<llvm::Win64EH::RuntimeFunction>{
+      reinterpret_cast<const llvm::Win64EH::RuntimeFunction*>(exception_table_bytes.data()),
+      exception_table_bytes.size() / sizeof(llvm::Win64EH::RuntimeFunction)};
+}
+
+ErrorMessageOr<const llvm::Win64EH::RuntimeFunction*>
+orbit_object_utils::CoffFileImpl::GetPrimaryRuntimeFunction(
+    const llvm::Win64EH::RuntimeFunction* runtime_function) {
+  const llvm::Win64EH::UnwindInfo* unwind_info{};
+  while (true) {
+    uintptr_t unwind_info_bytes{};
+    if (object_file_->getRvaPtr(runtime_function->UnwindInfoOffset, unwind_info_bytes)) {
+      return ErrorMessage{absl::StrFormat("Unable to read RUNTIME_FUNCTION at RVA %#x.",
+                                          runtime_function->UnwindInfoOffset)};
+    }
+
+    unwind_info = reinterpret_cast<const llvm::Win64EH::UnwindInfo*>(unwind_info_bytes);
+    // "UNW_FLAG_CHAININFO: This unwind info structure is not the primary one for the procedure.
+    // Instead, the chained unwind info entry is the contents of a previous RUNTIME_FUNCTION entry."
+    if ((unwind_info->getFlags() & llvm::Win64EH::UNW_ChainInfo) == 0) {
+      break;
+    }
+
+    runtime_function = unwind_info->getChainedFunctionEntry();
+  }
+
+  return runtime_function;
+}
+
+ErrorMessageOr<std::vector<CoffFileImpl::UnwindRange>> CoffFileImpl::GetUnwindRanges() {
+  auto runtime_functions_or_error = GetRuntimeFunctions();
+  if (runtime_functions_or_error.has_error()) {
+    return ErrorMessage{absl::StrFormat("Unable to load unwind info ranges: %s",
+                                        runtime_functions_or_error.error().message())};
+  }
+  const llvm::ArrayRef<llvm::Win64EH::RuntimeFunction>& runtime_functions =
+      runtime_functions_or_error.value();
+
+  std::vector<UnwindRange> unwind_ranges;
+  // For a function, the primary RUNTIME_FUNCTION and the ones with UNW_FLAG_CHAININFO set (if any)
```

Comment:
Do I understand correctly here that if there are multiple chained functions that all have the same primary function, we will always end up with symbols that can be ordered S1, S2, S3, ..., SN such that the address range of Si is contained in Si+1? All have the same start address (of the primary runtime function) and we always take the entire range from start to end. Do we remove these "subranges" later? Could there be any problems because of this?

Reply:
Ok, got it.